### DISCUSSION TITLE
Here's a little question.

### DISCUSSION THREAD

**Discussion by jamie2minchae** [bfieldtools/bfieldtools#46] [Q&A] [answered]
Hi, I'm a graduate student in Korea, just starting coding with python.
I always use bfieldtools thankfully, and my goal is to make homogeneous magnetic field by combining spherical harmonics.
Today, my question is about creating a blender-based mesh file.
I'm going to make the 10X10_plane_hires.obj file that you use as an exercise in a blender.
Next, I'm going to make a circular plate divided into trimesh.

However, when I tried to divide the mesh on the blender, there was a problem that it was divided into rectangular meshes as below, not the desired mesh...
![image](https://user-images.githubusercontent.com/108649323/177094541-20225006-4007-40d0-8a1a-4cc1e70b552d.png)

I experienced github for the first time by bfieldtools. 
I'm worried that this posting of the question would be a nuisance to you.
Thanks a lot for taking the time to read the long question!!

**Comment by rzetter** [comment, marked as answer]
Hi Jamie! Don't worry,  you're not being a nuisance at all. 

This is a good question. The underlying "issue" here is that the way Blender creates meshes is not optimized for physics modelling like FEM, BEM or bfieldtools. In these physics modelling applications, we would like to have equilateral-ish triangles whose size is somehow related to the geometry of the mesh, e.g. the triangles may be smaller where the mesh is thinner or has a hole etc. The mesh you generated in Blender **should work** as long as it is a triangle mesh (rather than quad mesh), but it won't be quite as fit-for-purpose. If you want, you can simply create a higher-resolution mesh and take a performance/memory use hit in order to compensate for the sub-optimal meshing. 

There are lots of different tools specifically meant for physics-modelling mesh generation. One of the most(?) popular open-source ones is [gmsh](https://gmsh.info/). However, I found gmsh to be somewhat hard to use. If you have access to COMSOL at your university, that software has a pretty easy-to-use meshing functionality as well as pretty comprehensive tutorials!

**Reply by jamie2minchae** [reply]
Thank you so much for your kind reply!

Referring to your comment, I optimized the mesh on COMSOL, and exported it as a text file.
Afterwards, I matched the vertex and face information of the mesh extracted as a .txt file to the extension format of the .obj file.
So the problem was solved well!

If one more question is possible, I would like to know why you created a sample area by adjusting the blm[i] -coefficient instead of specifying the l,m value for Ylm at the example of "High order spherical harmonic biplanar coil design".
This question came up for the following reasons.
I want to combine the real spherical harmonic functions by specifying Ylm at the cylindrical sample area,
but it's difficult to know which Ylm functions are designated if I specify the blm coefficient.

Respectfully yours,
Jamie

**Reply by rzetter** [reply]
Hi!

Regarding exporting meshes from COMSOL, you should be able to directly export in OBJ or STL format rather than having to do it via a .txt conversion. There should be a file type dropwdown when exporting.

Regarding the spherical harmonics coefficients, have you looked at the bfieldtools papers? Specifically, check out Section 4A here: https://aip.scitation.org/doi/10.1063/5.0016090. As you can see, we split each harmonic function 𝑌𝑙𝑚 into two terms; The 𝛼𝑙𝑚 terms involve powers of the inverse distance, representing sources close to the origin, whereas the 𝛽𝑙𝑚 terms represent far-away sources. When designing coils, the terms that go to zero at the origin are typically the ones of interest, i.e. 𝛽𝑙𝑚. NB! The radius at which the 𝛼𝑙𝑚 and 𝛽𝑙𝑚 cross over can be customized in the sphtools functions as needed. The vector blm[i] in that example is simply a flattened version of a 𝛽[𝑙,𝑚] array. The conversion from l,m to i is done as below:

```
L, M = 1, -1 --- i = 0
L, M = 1,  0 --- i = 1
L, M = 1,  1 --- i = 2
L, M = 2, -2 --- i = 3
L, M = 2, -1 --- i = 4
L, M = 2,  0 --- i = 5
L, M = 2,  1 --- i = 6
L, M = 2,  2 --- i = 7
L, M = 3, -3 --- i = 8
L, M = 3, -2 --- i = 9
L, M = 3, -1 --- i = 10
L, M = 3,  0 --- i = 11
```

Also, note that the example you point to only uses spherical harmonics to define the desired field, but does not use the spherical harmonics framework when optimizing the coils. For a more elegant and computationally efficient example, see [this one](https://bfieldtools.github.io/auto_examples/coil_design/spherical_harmonics_coil_design.html#sphx-glr-auto-examples-coil-design-spherical-harmonics-coil-design-py). However, this linked example obviously is limited to spherical regions of interest rather than e.g. cylindrical.

**Reply by jamie2minchae** [reply]
Hello, I really appreciate you for your kind and helpful advice!

I think I should continue to study while reading your excellent papers.

Once my research is completed, I would like to thank you again. Have a nice day!
